lib/Mitigation.pm

```diff
@@ -0,0 +1,335 @@
+use ipmi_backend_utils;
+use power_action_utils 'power_action';
+
+
+=head2 reboot_and_wait
+
+	reboot_and_wait([timeout => $timeout]);
+
+To reboot and waiting system back and login it.
+This could support IPMI and QEMU backend.
+C<$timeout> in seconds.
+=cut
+sub reboot_and_wait {
+    my ($self, $timeout) = @_;
+    power_action('reboot', textmode => 1, keepconsole => 1);
+    if (check_var('BACKEND', 'ipmi')) {
+        switch_from_ssh_to_sol_console(reset_console_flag => 'on');
+        check_screen('login_screen', $timeout);
+        use_ssh_serial_console;
+    }
+    else {
+        $self->wait_boot(textmode => 1, ready_time => 300,, in_grub => 1);
+        select_console 'root-console';
+    }
+}
+
+our $syspath = '/sys/devices/system/cpu/vulnerabilities/';
+
+#If you base on this Class, you have to override new function,
+#so that it could be loaded by openQA.
+#
+sub new {
+    my ($class, $args) = @_;
+    return bless $args, $class;
+}
+
+sub Name {
+    my ($self, $value) = @_;
+    if (@_ == 2) {
+        $self->{name} = $value;
+    }
+    return $self->{name};
+}
+
+sub Parameter {
+    my ($self, $value) = @_;
+    if (@_ == 2) {
+        $self->{parameter} = $value;
+    }
+    return $self->{Parameter};
+}
+sub sysfs_name {
+    my ($self, $value) = @_;
+    if (@_ == 2) {
+        $self->{sysfs_name} = $value;
+
+    }
+    return $self->{sysfs_name};
+}
+sub CPUID {
+    my ($self, $value) = @_;
+    return $self->{CPUID};
+}
+
+sub MSR {
+    my ($self, $value) = @_;
+    return $self->{IA32_ARCH_CAPABILITIES};
+}
+
+sub read_cpuid {
+    my $self = shift;
+    zypper_call('in cpuid');
+    my $edx = hex script_output(
+        "cpuid -1 -l 7 -s 0 -r | awk \'{print \$6}\' | awk -F \"=\" \'{print \$2}\' | tail -n1"
+    );
+    return $edx;
+}
+
+sub read_msr {
+    my $self = shift;
+    script_output('modprobe msr');
+    my $edx = script_output(
+        "perl -e \'open(M,\"<\",\"/dev/cpu/0/msr\") and seek(M,0x10a,0) and read(M,\$_,8) and print\' | od -t u8 -A n"
+    );
+    return $edx;
+}
+
+
+sub show {
+    my $self = shift;
+    print $self->Name(),  ",";
+    print $self->CPUID(), ",";
+    print $self->MSR(),   "\n";
+}
+
+sub vulnerabilities {
+    my $self = shift;
+    if ($self->read_cpuid() & $self->CPUID()) {
+        if ($self->read_msr() & $self->MSR()) {
+            record_info("Not Affected", "This machine needn't be tested.");
+            return 0;    #Not Affected
+        }
+    }
+    record_info("vulnerable", "Testing will continue.");
+    return 1;            #Affected
+}
+
+sub sysfs {
+    my ($self, $value) = @_;
+    $value =~ s/,/_/g;
+    if (@_ == 2) {
+        return $self->{sysfs}->{$value};
+    }
+    return $self->{sysfs};
+
+}
+
+sub dmesg {
+    my $self = shift;
+    for my $p (keys %{$self->{dmesg}}) {
+        print "dmesg ", $self->Name, "\n";
+        print $self->{dmesg}->{$p}, "\n";
+    }
```

Comment:
If you want to connect two strings, please use `.`   Same question in other places

Reply:
OK. Thanks.